File: leetcode/01.01.cpp
#include <string>
#include <unordered_set>
#include <algorithm>

using namespace std;

class Solution {
public:
  bool isUnique(string astr) {
    unordered_set<char> cache;
    for (char c : astr) {
      if (cache.find(c) != cache.end()) {
        return false;
      }
      cache.emplace(c);
    }
    return true;
  }

  bool isUnique1(string astr) {
    sort(astr.begin(), astr.end());
    int n = astr.size();
    for (int i = 1; i < n; ++i) {
      if (astr[i] == astr[i - 1]) {
        return false;
      }
    }
    return true;
  }

  bool isUnique2(string astr) {
    int mask = 0;
    for (char c : astr) {
      int val = 1 << (c - 'a');
      if ((val & mask) != 0) {
        return false;
      }
      mask |= val;
    }
    return true;
  }
};

File: leetcode/01.01.py
class Solution:

    def isUnique(self, astr: str) -> bool:
        cache = set()
        for c in astr:
            if c in cache:
                return False
            cache.add(c)
        return True

    def isUnique1(self, astr: str) -> bool:
        return len(astr) == len(set(astr))

    def isUnique2(self, astr: str) -> bool:
        mask = 0
        for c in astr:
            val = 1 << (ord(c) - ord('a'))
            if mask & val != 0:
                return False
            mask = mask | val
        return True